File: 107_word-break/word-break.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/word-break
@Language: Python
@Datetime: 16-06-02 09:10
'''

class Solution:
    # @param s: A string s
    # @param dict: A dictionary of words dict
    def wordBreak(self, s, dict):
        # write your code here
        #corner case
        if len(dict) == 0:
            return len(s) == 0
        n = len(s)
        maxLength = max([len(w) for w in dict])
        f = [False] * (n + 1)
        f[0] = True
        for i in range(1, n + 1):
            for j in range(1, min(i, maxLength) + 1):
                if not f[i - j]:
                    continue
                if s[i - j:i] in dict:
                    f[i] = True
                    break
        return f[-1]
        

File: 107_word-break/word-break.java
/*
@Copyright:LintCode
@Author:   hanqiao
@Problem:  http://www.lintcode.com/problem/word-break
@Language: Java
@Datetime: 16-06-04 03:37
*/

public class Solution {
    /**
     * @param s: A string s
     * @param dict: A dictionary of words dict
     */
    private int findMaxLength(Set<String> dict) {
        int max = 0;
        for (String s : dict) {
            if (s.length() > max) {
                max = s.length();
            }
        }
        return max;
    }
    
    public boolean wordBreak(String s, Set<String> dict) {
        // write your code here
        if (dict.size() == 0) {
            return s.length() == 0;
        }
        
        int n = s.length();
        int maxLength = findMaxLength(dict);
        boolean[] f = new boolean[n + 1];
        f[0] = true;
        
        for (int i = 1; i < n + 1; i++) {
            f[i] = false;
            for (int j = 1; j < Math.min(i, maxLength) + 1; j++) {
                if (!f[i - j]) {
                    continue;
                }
                if (dict.contains(s.substring(i - j, i))) {
                    f[i] = true;
                    break;
                }
            }
        }
        return f[n];
    }
}